deeppavlov/models/ranking/tf_base_matching_model.py

```diff
@@ -0,0 +1,192 @@
+
+        Args:
+            batch (List[List[np.ndarray]]): List of samples with model inputs each:
+                [( context, context_len, response, response_len ), ( ... ), ... ].
+        Returns:
+            Dict: feed_dict to feed a model
+        """
+        input_context = []
+        input_context_len = []
+        input_response = []
+        input_response_len = []
+
+        # format model inputs as numpy arrays
+        for sample in batch:
+            input_context.append(sample[0])
+            input_context_len.append(sample[1])
+            input_response.append(sample[2])
+            input_response_len.append(sample[3])
+
+        return {
+            self.utterance_ph: np.array(input_context),
+            self.all_utterance_len_ph: np.array(input_context_len),
+            self.response_ph: np.array(input_response),
+            self.response_len_ph: np.array(input_response_len)
+        }
+
+    def _predict_on_batch(self, batch: Dict) -> np.ndarray:
+        """
+        Run a model with the batch of inputs.
+        The function returns a list of predictions for the batch in numpy format
+
+        Args:
+            batch (Dict): feed_dict that contains a batch with inputs for a model
+
+        Returns:
+            nd.array: predictions for the batch
+        """
+        return self.sess.run(self.y_pred, feed_dict=batch)[:, 1]
+
+    def _train_on_batch(self, batch: Dict, y: List[int]) -> float:
+        """
+        The function is for formatting of feed_dict used as an input for a model
+        Args:
+            batch (Dict): feed_dict that contains a batch with inputs for a model (except ground truth labels)
+            y (List(int)): list of ground truth labels
+
+        Returns:
+            float: value of mean loss on the batch
+        """
+        batch.update({self.y_true: np.array(y)})
+        return self.sess.run([self.loss, self.train_op], feed_dict=batch)[0]  # return the first item aka loss
+
+    # load() and save() are inherited from TFModel
+
+    def train_on_batch(self, samples_generator: Iterable[List[np.ndarray]], y: List[int]) -> float:
+        """
+        This method is called by trainer to make one training step on one batch.
+
+        Args:
+            samples_generator (Iterable[List[np.ndarray]]): generator that returns list of numpy arrays
+            of words of all sentences represented as integers.
+            Has shape: (number_of_context_turns + 1, max_number_of_words_in_a_sentence)
+            y (List[int]): tuple of labels, with shape: (batch_size, )
+
+        Returns:
+            float: value of mean loss on the batch
+        """
+        loss = 0
+        buf = []
+        j = 0
+        while True:
+            try:
+                sample = next(samples_generator)
+                j += 1
+                self._append_sample_to_batch_buffer(sample=sample, buf=buf)
+                if len(buf) >= self.batch_size:
```

Comment:
The length of  the `list(samples_generator)` here is always equal to the `self.batch_size`. You could simply convert the samples_generator to the list and use the `_make_batch` function to process it.

Reply:
It still needs to call `self.__append_sample_to_batch_buffer` function, e.g. when dealing with inputs of the SMN/DAM/DAM-USE-T models.
So I write the code that is equal to `list(batch)` but uses `self.__append_sample_to_batch_buffer` to format `buf` as a `List[Tuple[np.ndarray]]`.
Please, see this:
https://github.com/deepmipt/DeepPavlov/blob/ranking_matching_models_use_transformer/deeppavlov/models/ranking/siamese_model.py#L47

Reply:
ok.